### PR TITLE
Feature/peer 39 get user list

### PR DESCRIPTION
**Pull request by SharyuMarwadi**

### CODE REVIEW

File: internal/api/user.go [hunk 1]
```diff
@@ -134,3 +135,34 @@ func registerUser(userSvc user.Service) http.HandlerFunc {
...
+			dto.ErrorRepsonse(rw, err)
+			return
+		}
+		pageInt, _ := strconv.Atoi(page)
+		perPage := req.URL.Query().Get("page_size")
+		var perPageInt int
+		if perPage == "" {
+			perPageInt = constants.DefaultPageSize
+		} else {
+			perPageInt, _ = strconv.Atoi(perPage)
```
Comment (yama11299) [inline internal/api/user.go:153]: What if one passed negative page size?

File: internal/api/user_test.go [hunk 2]
```diff
@@ -310,3 +310,81 @@ func TestLoginUser(t *testing.T) {
...
+		expectedStatusCode int
+	}{
+		{
+			name:      "Success for get user list",
+			authToken: "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjo2fQ.XaYo0qdBCdDh1-nEeuUSdTbtp0enWFIySKnw-oQpTBg",
+			page:      "1",
+			per_page:  "10",
+			paramName: "sharyu%20marwadi",
+			setup: func(mockSvc *mocks.Service) {
+				mockSvc.On("ListUsers", mock.Anything, mock.Anything).Return(dto.UserListWithMetadata{}, nil).Once()
```
Comment (yama11299) [inline internal/api/user_test.go:334]: why mock.Anything as 2nd arg and why result is empty in success scenario?

File: internal/api/user_test.go [hunk 3]
```diff
@@ -310,3 +310,81 @@ func TestLoginUser(t *testing.T) {
...
+			expectedStatusCode: http.StatusOK,
+		},
+		{
+			name:      "Internal server error",
+			authToken: "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjo2fQ.XaYo0qdBCdDh1-nEeuUSdTbtp0enWFIySKnw-oQpTBg",
+			page:      "1",
+			per_page:  "10",
+			paramName: "sharyu%20marwadi",
+			setup: func(mockSvc *mocks.Service) {
+				mockSvc.On("ListUsers", mock.Anything, mock.Anything).Return(dto.UserListWithMetadata{}, apperrors.InternalServerError).Once()
```
Comment (yama11299) [inline internal/api/user_test.go:345]: remove mock.Anything for 2nd arg

File: internal/app/users/service_test.go [hunk 4]
```diff
@@ -328,3 +343,61 @@ func TestLoginUser(t *testing.T) {
...
+		setup           func(userMock *mocks.UserStorer)
+		isErrorExpected bool
+	}{
+		{
+			name:    "Success for get user list",
+			context: context.Background(),
+			reqData: dto.UserListReq{},
+			setup: func(userMock *mocks.UserStorer) {
+				userMock.On("GetTotalUserCount", mock.Anything, mock.Anything).Return(int64(280), nil).Once()
+				userMock.On("ListUsers", mock.Anything, mock.Anything).Return([]repository.User{}, nil).Once()
```
Comment (yama11299) [inline internal/app/users/service_test.go]: same comment as given above

File: internal/pkg/dto/users.go [hunk 5]
```diff
@@ -74,6 +74,30 @@ type GetUserListReq struct {
 	Page      int64
 }
 
+type UserListResp struct {
```
Comment (yama11299) [inline internal/pkg/dto/users.go:77]: rename to User.

File: internal/pkg/dto/users.go [hunk 6]
```diff
@@ -74,6 +74,30 @@ type GetUserListReq struct {
...
+	Id        int64  `json:"id"`
+	Email     string `json:"email"`
+	FirstName string `json:"first_name"`
+	LastName  string `json:"last_name"`
+}
+
+type UserListMetadata struct {
```
Comment (yama11299) [inline internal/pkg/dto/users.go:84]: rename to PageToken

File: internal/pkg/dto/users.go [hunk 7]
```diff
@@ -74,6 +74,30 @@ type GetUserListReq struct {
...
+	TotalCount  int64 `json:"total_count"`
+	CurrentPage int64 `json:"page"`
+	PageCount   int64 `json:"page_size"`
+}
+
+type UserListWithMetadata struct {
```
Comment (yama11299) [inline internal/pkg/dto/users.go:90]: rename to ListUserResponse.

File: internal/repository/postgresdb/user.go [hunk 8]
```diff
@@ -189,4 +190,75 @@ func (us *userStore) SyncData(ctx context.Context, updateData dto.User) (err err
 
 	return
 
-}
\ No newline at end of file
+}
+
+func (us *userStore) GetTotalUserCount(ctx context.Context, reqData dto.UserListReq) (totalCount int64, err error) {
+
+	queryBuilder := repository.Sq.Select("count(*)").From("users")
```
Comment (yama11299) [inline internal/repository/postgresdb/user.go:197]: use table const from method receiver.

**Comment by SharyuMarwadi** [comment]
Review changes don

File: internal/repository/postgresdb/user.go [hunk 9]
```diff
@@ -243,8 +240,7 @@ func (us *userStore) ListUsers(ctx context.Context, reqData dto.UserListReq) (re
 
 	listUsersQuery, args, err := queryBuilder.ToSql()
 	if err != nil {
-		logger.Errorf("error in generating squirrel query, err: %s", err.Error())
-		err = apperrors.InternalServerError
+		err = fmt.Errorf("error in generating squirrel query, err: %w", err)
```
Comment (yama11299) [inline internal/repository/postgresdb/user.go:243]: remove words like squirrel if used anywhere else also

Comment (yama11299) [reply internal/repository/postgresdb/user.go:243]: Suggestions:
"falied to create query" 
"failed to execute query"

File: internal/api/user.go [hunk 10]
```diff
@@ -173,12 +173,20 @@ func listUsersHandler(userSvc user.Service) http.HandlerFunc {
 			dto.ErrorRepsonse(rw, err)
 			return
 		}
+
 		names := strings.Split(req.URL.Query().Get("name"), " ")
+		self := req.URL.Query().Get("exclude_self")
+		selfBool := false
+		if self == "true" {
```
Comment (yama11299) [inline internal/api/user.go:180]: what if we get True or TRUE?

Comment (yama11299) [reply internal/api/user.go:180]: selfBool, err := strconv.ParseBool(self)

File: internal/app/users/service.go [hunk 11]
```diff
@@ -303,11 +303,24 @@ func (us *service) ListUsers(ctx context.Context, reqData dto.ListUsersReq) (res
...
+				users = append(users, user)
+			}
+		} else {
+			user := mapDbUserToUserListResp(dbUser)
+			users = append(users, user)
+		}
+	}
+
+	if reqData.Self && totalCount > 0 {
+		totalCount = totalCount - 1
```
Comment (yama11299) [inline internal/app/users/service.go:323]: What is this logic for?

File: internal/pkg/constants/system.go [hunk 12]
```diff
@@ -1,5 +1,8 @@
 package constants
 
+type UserIdCtxKey string
+type RoleCtxKey string
```
Comment (yama11299) [inline internal/pkg/constants/system.go:4]: don't define type for each ctx key, simply declare type ContextKey string and use it for all keys

File: internal/app/users/service.go (hunk 11 above)
Comment (SharyuMarwadi) [reply internal/app/users/service.go:323]: If the logged in person is excluded from the user list, the total count must be reduced by one

Comment (yama11299) [reply internal/app/users/service.go:323]: sorry my bad, actually I didn't select all the lines for which I was asking. check the suggestion

File: internal/app/users/service.go [hunk 13]
```diff
@@ -303,11 +303,24 @@ func (us *service) ListUsers(ctx context.Context, reqData dto.ListUsersReq) (res
...
+		if reqData.Self {
+			if dbUser.Id != userId {
+				user := mapDbUserToUserListResp(dbUser)
+				users = append(users, user)
+			}
+		} else {
+			user := mapDbUserToUserListResp(dbUser)
+			users = append(users, user)
+		}
+	}
```
Comment (yama11299) [inline internal/app/users/service.go:320]: if reqData.Self && dbUser.Id == userID {
    continue
}

user := mapDbUserToUserListResp(dbUser)
users = append(users, user)

This should work :)

**Comment by SharyuMarwadi** [comment]
Changes done

**Pull request closed without merging**